arch/arm/core/aarch32/Kconfig

```diff
@@ -76,6 +76,14 @@ config ISA_ARM
 	  processor start-up. Much of its functionality was subsumed into T32 with
 	  the introduction of Thumb-2 technology.
 
+config COMPILER_ISA_THUMB2
+	bool "Compile C functions using Thumb-2 instruction set"
+	depends on ISA_THUMB2
+	default y
+	help
+	  This option configures the compiler to compile all C functions using
+	  the Thumb-2 instruction set.
+
```

Comment:
Let's have 2 config options.

C_THUMB2
ASM_THUMB2

ASM determines if thumb is used for assembly.
C      determines if thumb is used for C.

This would make it more obvious that using thumb for C doesn't imply
using it for ASM and vice-versa.

Reply:
```
2. When `ISA_ARM` and `ISA_THUMB2` symbols are selected simultaneously
  (i.e. on Cortex-A and Cortex-R processors), make all ASM functions
  use the ARM instruction set by default.
```

Not having a separate symbol for the ASM target ISA selection and prioritising/forcing ARM over Thumb-2 for ASM on the processors that support both ISAs was a deliberate decision made due to the following reasons:

1. The Cortex-R ASM code in Zephyr is currently written using the ARM ISA mnemonics and cannot be assembled into the Thumb-2 ISA instructions (prior to this PR, `ISA_THUMB2` was not selected on the Cortex-R, so all ASM code was being assembled into the ARM instructions).
2. The "default state" for the cores that support both ARM and Thumb-2 ISAs (i.e. Cortex-A and Cortex-R family cores) is the ARM. For example, when the processor executes an exception vector (which is specified and implemented in ASM), it assumes the ARM mode and the exception vector jump instruction must be encoded using the ARM ISA.

IMHO, `COMPILER_ISA_THUMB2` is an adequate name for the symbol for the following reasons:

1. This symbol is intended to specify Thumb-2 as the target ISA for all "compiled" code (since Zephyr only supports C/C++ at the this time, it only refers to the C/C++ language compilations but is not necessarily limited to those languages; I will revise the symbol description to be more concise about this).
2. The scope of `COMPILER_ISA_THUMB2` is limited to the "compiled" code (and not ASM) because, in conventional sense, ASM code is "assembled" by an "assembler" and not "compiled" by a "compiler" (for instance, an equivalent symbol for the ASM code would be named `ASSEMBLER_ISA_THUMB2`).

Reply:
```
When `ISA_ARM` and `ISA_THUMB2` symbols are selected simultaneously
  (i.e. on Cortex-A and Cortex-R processors), make all ASM functions
  use the ARM instruction set by default.
```

I want this to be expressed in Kconfig instead of in preprocessor ifdefs. Because then it will be clearer that these two options are not always set the same and it will be easier to understand
what determines if they are set or not.

e.g.

```
config ASSEMBLER_ISA_THUMB2
  bool
  default y if  CONFIG_ISA_THUMB2
  depends on ! CONFIG_ISA_ARM
```

As far as I understand the semantics I agree with them, I just want them expressed in a way that will be faster to understand.

Reply:
Using COMPILER and ASSEMBLER instead of C and ASM is OK.

Reply:
While introducing `ASSEMBLER_ISA_THUMB2` and defining constraints there could be somewhat helpful, in the long run, I think it will create unnecessary misunderstanding and confusion, as this is not really a configurable/selectable option (it can give developers the false impression that they have the option to choose the target ISA for the ASM code, when they really do not; at least, not globally).

To elaborate:
* For the cores that only support the Thumb(-2) ISA (i.e. Cortex-M family), `ASSEMBLER_ISA_THUMB2` **must be selected** (to do otherwise simply does not make sense as there is no other ISA supported).
* For the cores that support both the ARM and Thumb-2 ISA (i.e. Cortex-A and Cortex-R family), once again, `ASSEMBLER_ISA_THUMB2` **must be selected**; otherwise, the assembler will output errors because of the ARM instructions being used (note that the current Cortex-R, and to-be Cortex-A as well, ASM code is implemented using the ARM ISA, so `ASSEMBLER_ISA_THUMB2` must always be selected).

As I mentioned in my previous comment, the "default" ISA for the cores that support both the ARM and Thumb-2 ISAs is the ARM ISA, and there are both technical (exception vectors) and practical (ARM instructions are more semantically packed and almost always result in a lower instruction count) reasons as to why the ARM ISA for ASM code is implied when the ARM ISA support is specified; **hence, this should not be/should not be deemed something that is configurable.**

Reply:
I agree that having a configurable/selectable option would be a bad idea, but it won't be configurable/selectable because it will be promptless option.

Reply:
Hi, did you accidentally contradict yourself, or am I misunderstanding something, I don't understand how both

> When `ISA_ARM` and `ISA_THUMB2` symbols are selected simultaneously
>   (i.e. on Cortex-A and Cortex-R processors), make all ASM functions
>   use the ARM instruction set by default.
> 

and

> For the cores that support both the ARM and Thumb-2 ISA (i.e. Cortex-A and Cortex-R family), once again, ASSEMBLER_ISA_THUMB2 must be selected

can be true.

Reply:
> Hi, did you accidentally contradict yourself, or am I misunderstanding something, I don't understand how both
> 
> > When `ISA_ARM` and `ISA_THUMB2` symbols are selected simultaneously
> > (i.e. on Cortex-A and Cortex-R processors), make all ASM functions
> > use the ARM instruction set by default.
> 
> and
> 
> > For the cores that support both the ARM and Thumb-2 ISA (i.e. Cortex-A and Cortex-R family), once again, ASSEMBLER_ISA_THUMB2 must be selected
> 
> can be true.

@SebastianBoe I did accidentally contradict myself. 

It should read:
> For the cores that support both the ARM and Thumb-2 ISA (i.e. Cortex-A and Cortex-R family), ~once again~, ASSEMBLER_ISA_THUMB2 must **NOT** be selected; otherwise, the assembler will output errors because of the ARM instructions being used (note that the current Cortex-R, and to-be Cortex-A as well, ASM code is implemented using the ARM ISA, so ASSEMBLER_ISA_THUMB2 must always be **NOT** selected).

It looks like lack of sleep is taking its toll 👎 

Either way, my point that this symbol is not and should not be configurable holds.

Reply:
> I agree that having a configurable/selectable option would be a bad idea, but it won't be configurable/selectable because it will be a promptless option.

I will look into adding `ASSEMBLER_ISA_THUMB2` tomorrow (with the emphasis that this symbol is solely there as a helper symbol).

Reply:
Done